board/board.h

```diff
@@ -0,0 +1,57 @@
+// ///////////////////////////////////////////////////////////// //
+// Hardware abstraction layer for all different supported boards //
+// ///////////////////////////////////////////////////////////// //
+#include "board_declarations.h"
+#include "boards/common.h"
+
+// ///// Board definition and detection ///// //
+#include "drivers/harness.h"
+#ifdef PANDA
+  #include "boards/white.h"
+  #include "boards/grey.h"
+  #include "boards/black.h"
+#else
+  #include "boards/pedal.h"
+#endif
+
+void detect_board_type(void){
+  #ifdef PANDA
+    // SPI lines floating: white (TODO: is this reliable?)
+    if((detect_with_pull(GPIOA, 4, PULL_DOWN)) || (detect_with_pull(GPIOA, 5, PULL_DOWN)) || (detect_with_pull(GPIOA, 6, PULL_DOWN)) || (detect_with_pull(GPIOA, 7, PULL_DOWN))){
+      hw_type = HW_TYPE_WHITE_PANDA;
+      current_board = &board_white;
+    } else if(detect_with_pull(GPIOA, 13, PULL_DOWN)) { // Rev AB deprecated, so no pullup means black. In REV C, A13 is pulled up to 5V with a 10K
+      hw_type = HW_TYPE_GREY_PANDA;
+      current_board = &board_grey;
+    } else {
+      hw_type = HW_TYPE_BLACK_PANDA;
+      current_board = &board_black;
+    }
+  #else
```

Comment:
can be simplified in:
```
#elseif PEDAL
  hw_type = HW_TYPE_PEDAL;
  current_board = &board_pedal;
#else 
    hw_type = HW_TYPE_UNKNOWN;
    puts("Hardware type is UNKNOWN!\n");
 #endif

Reply:
Also, if the type is UNKNOWN, does it make sense to keep executing code? i.e. are we sure we won't have unexpected consequences?

Reply:
If the type is UNKNOWN, it stops executing in the main loop :)
```
// check for non-supported board types
  if(hw_type == HW_TYPE_UNKNOWN){
    puts("Unsupported board type\n");
    while (1) { /* hang */ }
  }
```

Reply:
Nice!